timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,954 @@
+                # if either atom is tetrahedral then it has no pi bonds to conjugate over
+                # exception: weird sulfur groups
+                continue
+            else:
+                canonical_chiral_bonds.add(k)
+
+    return canonical_chiral_bonds
+
+
+def find_chiral_atoms(mol):
+    """
+    Find chiral atoms in a molecule. Chirality is perceived locally, without taking
+    graph-symmetry into account. Eg. even something like CH4 would be considered
+    to be chiral since the hydrogens are distinguishable.
+
+    Returns
+    -------
+    set
+        Set of chiral atoms.
+
+    """
+    mol_geom = geometry.classify_geometry(mol)
+    chiral_atoms = set()
+    for a in mol.GetAtoms():
+        a_idx = a.GetIdx()
+        nbs = a.GetNeighbors()
+        if len(nbs) == 4:
+            chiral_atoms.add(a_idx)
+        elif len(nbs) == 3 and mol_geom[a_idx] == LocalGeometry.G3_PYRAMIDAL:
+            # Trivalent Sulfur/Phosphorus typically are not invertible, nor
+            # are they properly parameterized. So we assert if we detect this case.
+            if a.GetAtomicNum() == 16 or a.GetAtomicNum() == 15:
+                assert 0, "Pyramidal Phosphorous or Sulfur detected."
+        elif len(nbs) > 4:
+            assert 0, "More than four neighbors found"
+    return chiral_atoms
+
+
+def check_bond_stability(j, k, bond_idxs, bond_params):
+    # tbd this should really be based on the pdf, we should expect to see
+    # near zero probability at bond lengths < 0.5 Angstroms
+    canonical_bonds = dict()
+    for idxs, params in zip(bond_idxs, bond_params):
+        canonical_bonds[dummy.canonicalize_bond(idxs)] = params
+
+    jk = dummy.canonicalize_bond((j, k))
+
+    if jk not in canonical_bonds:
+        return False
+    # these are lower bound (conservative estimates) for what's required
+    # to prevent the bond from sampling a length of 0 or from drifting
+    # arbitrarily apart. Angles defined over these bonds should hopefully
+    # then be stable.
+    elif canonical_bonds[jk][0] < 50.0:
+        return False
+    elif canonical_bonds[jk][1] < 0.05:
+        return False
+
+    return True
+
+
+def check_angle_stability(j, k, l, angle_idxs, angle_params):
+
+    canonical_angles = dict()
+    for idxs, params in zip(angle_idxs, angle_params):
+        force_const, theta = params
+        assert force_const > 0.0
+        assert theta >= 0.0
+        assert theta <= np.pi
+        canonical_angles[dummy.canonicalize_bond(idxs)] = params
+
+    jkl = dummy.canonicalize_bond((j, k, l))
+
+    # similar to bonds, these hopefully will prevent angles from linearizing
+    # eg. in the case of nitrile groups or from spinning arbitrarily. these
+    # angle terms are typically used to define cross-products.
+    if jkl not in canonical_angles:
+        return False
+    elif canonical_angles[jkl][0] < 50.0:
+        return False
+    elif abs(canonical_angles[jkl][1] - 3.1415) < 0.05:
```

Comment:
Should we use `np.pi` here, or is there a reason to prefer 3.1415?

Reply:
willfix to np.pi